### ISSUE TITLE
clipping of adversarial examples missing in PGD attack

### ISSUE THREAD

**Issue by fahad92virgo** [IBM/adversarial-robustness-toolbox#148]
**Describe the bug**
For the projected gradient descent attack, we need to perform two projections.
1) ensure that the perturbation do not exceed the maximum allowed perturbation
2) ensure that the resulting adversarial example is in the valid image range. 

In the current implementation, the second projection is missing. I was getting very different results for the robustness accuracy to a PGD attack when using ART or the code from the TRADES paper. The resulting adversarial examples do not conform to the clip values set in the PyTorch classifier

Upon looking at the code, I found that the projection step to a valid image is missing. 
In the original code from madry's lab, this projection can be seen here: https://github.com/MadryLab/cifar10_challenge/blob/master/pgd_attack.py#L59
In the code from the TRADES paper, the corresponding projection for an image normalized between 0 and 1 can be seen here: https://github.com/yaodongyu/TRADES/blob/master/pgd_attack_cifar10.py#L78

In the following ART's code, https://github.com/IBM/adversarial-robustness-toolbox/blob/master/art/attacks/fast_gradient.py#L273

I believe the following addition should be made.
```            
if project:
    perturbation = projection(x_adv[batch_index_1:batch_index_2] -x_init[batch_index_1:batch_index_2], eps, self.norm)
    x_adv[batch_index_1:batch_index_2] = x_init[batch_index_1:batch_index_2] + perturbation
    # addition
    if hasattr(self.classifier, 'clip_values') and self.classifier.clip_values is not None:
        clip_min, clip_max = self.classifier.clip_values
        x_adv[batch_index_1:batch_index_2] = np.clip(x_adv[batch_index_1:batch_index_2], clip_min, clip_max)
```

Note:
I am using art 0.10

**Comment by beat-buesser** [comment]
Hi @fahad92virgo Thank you for describing your observation. We will look at it as soon as possible. Would it be possible for you to share a short code section of how you are using PGD so that we can reproduce it?

**Comment by beat-buesser** [comment]
Hi @fahad92virgo I have compared PGD of ART and TRADES on the same datasets using pgd_attack_cifar10.py and pgd_attack_mnist.py of TRADES and I have obtained numerically identical results in terms of error and adversarial pixel values for MNIST and CIFAR10 using the default `eps` and `step_size` values of TRADES. In my tests all adversarial samples had all pixel values within the `clip_values` and I did not observe any pixel values outside of this valid image range.

As far as I can see it should not be necessary to add an additional clipping to line 277 because we are clipping the values to the image range with `_apply_perturbation` in line 271. Between lines 271 and 277 we are applying only the projection to `eps`.

To get exactly comparable results with PGD of ART and TRADES I have also tested turning off the random initialisation in both. It looks like that for TRADES the argument `--random=False` does not turn off random initialisation, I think because of missing conversion to type `bool`, therefore I have manually commented the lines 63-65 in the TRADES codes for MNIST and CIFAR10.

I would be very interested to see your code and to learn how you have created the described behaviour to make sure that we find all potential bugs. Until then I will close this issue.
